src/Symfony/Component/Mime/Crypto/DkimSigner.php

```diff
@@ -0,0 +1,215 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mime\Crypto;
+
+use Symfony\Component\Mime\Exception\InvalidArgumentException;
+use Symfony\Component\Mime\Exception\RuntimeException;
+use Symfony\Component\Mime\Header\UnstructuredHeader;
+use Symfony\Component\Mime\Message;
+use Symfony\Component\Mime\Part\AbstractPart;
+
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * RFC 6376 and 8301
+ */
+final class DkimSigner
+{
+    public const SIMPLE_CANON = 'simple';
+    public const RELAXED_CANON = 'relaxed';
+
+    public const SHA_256_ALGO = 'rsa-sha256';
+    // RFC 8463
+    public const ED25519_ALGO = 'ed25519-sha256';
+
+    private $key;
+    private $domainName;
+    private $selector;
+    private $defaultOptions;
+
+    /**
+     * @param string      $pk         The private key as a string or the path to the file containing the private key, should be prefixed with file:// (in PEM format)
+     * @param string|null $passphrase A passphrase of the private key (if any)
+     */
+    public function __construct(string $pk, string $domainName, string $selector, array $defaultOptions = [], string $passphrase = '')
+    {
+        if (!\extension_loaded('openssl')) {
+            throw new \LogicException('PHP extension "openssl" is required to use DKIM.');
+        }
+        if (!$this->key = openssl_pkey_get_private($pk, $passphrase)) {
+            throw new InvalidArgumentException('Unable to load DKIM private key: '.openssl_error_string());
+        }
+
+        $this->domainName = $domainName;
+        $this->selector = $selector;
+        $this->defaultOptions = $defaultOptions;
+    }
+
+    public function sign(Message $message, array $options = []): Message
+    {
+        $options += [
+            'algorithm' => self::SHA_256_ALGO,
+            'signature_expiration_delay' => 0,
+            'body_max_length' => PHP_INT_MAX,
+            'body_show_length' => false,
+            'header_canon' => self::RELAXED_CANON,
+            'body_canon' => self::RELAXED_CANON,
+            'headers_to_ignore' => [],
+        ] + $this->defaultOptions;
```

Comment:
shouldnt `defaultOptions` come first?

Reply:
Right, I've moved them in the constructor.